docs/developer/dependency_resolution.md

```diff
@@ -0,0 +1,190 @@
+        var service = new FooMaintenanceService(fooStore); //use the injected field
+    }
+}
+```
+
+The `@Inject` annotation on the `fooStore` field tells the extension loading mechanism that `FooMaintenanceExtension`
+depends on a `FooService` and because of that, any provider of a `FooStore` must be initialized _before_
+the `FooMaintenanceExtension`. The fact that `CosmosFooStoreExtension` provides a `FooStore` is declared using
+the `@Provides` annotation.
+
+### Option 2: use `@Requires` to declare dependencies
+
+In cases where defining a field seems unwieldy or is simply not desirable, we provide another way to dynamically resolve
+service from the context:
+
+```java
+
+@Requires({ FooService.class, /*maybe others*/ })
+public class FooMaintenanceExtension implements ServiceExtension {
+
+    @Override
+    public void initialize(ServiceExtensionContext context) {
+        var fooStore = context.getService(FooStore.class);
+        var service = new FooMaintenanceService(fooStore); //use the resolved object
+    }
+}
+```
+
+The important issue to mention is that `@Requires` is absolutely necessary to inform the service loader about the
+dependency. Failing to add it this may potentially result in exceptions, and in further consequence, in
+an `EdcInjectionException`.
+
+Option 1 and 2 are almost semantically equivalent, with the small exception of optional dependencies:
+while `@Inject(required=false)` allows for nullable dependencies, `@Requires` has no such option and the service
+dependency must be resolved explicitly using a boolean parameter `context.getService(FooStore.class, true)`.
+
+## Extension initialization sequence
+
+The extension loading mechanism uses a two-pass procedure to resolve dependencies. First, all implementors
+of `ServiceExtension` are instantiated using their public default constructor, put in a list and sorted using a
+topological sort algorithm based on their dependency graph. Cyclic dependencies would be reported in this stage.
+
+Second, the extension is initialized by setting all fields annotated with `@Inject` and by calling its `initialize()`
+method. This implies that every extension can assume that by the time its `initialize()` method executes, all its
+dependencies are already instantiated and registered, because the extension(s) providing them were ordered at previous
+positions in the list, and thus have already been initialized.
+
+## Limitations and differences to fully-fledged IoC containers
+
+#### Only available in `ServiceExtensions` (and why that's not entirely true)
+
+Services can only be injected into `ServiceExtension` objects at this time as they are the main hook points for plugins,
+and they have a clearly defined interface. All subsequent object creation must be done manually using conventional
+mechanisms like constructors or builders.
+
+**Well, actually...**
+
+However, using the `ReflectiveObjectFactory` together with an `Injector` and an `InjectionPointScanner` would
```

Comment:
I wonder if it's worth mentioning this at all. Wouldn't it be better to treat all these classes as internal?

Reply:
Yeah, probably better to remove and amend the document, if and when the injection mechanism gets expanded.